### ISSUE TITLE
Properly handle background thread processing in AndroidX tests

### ISSUE THREAD

**Issue by BenHenning** [oppia/oppia-android#89]
tl;dr: The project needs a way to consistently test operations which require background processing in a way that's consistent with production environments,

UI-centric AndroidX tests can run on both Robolectric and Espresso. Both Robolectric and Espresso have their own ways to block on background tasks completing.

Robolectric running in legacy mode can deterministically advance the system clock and looper to execute pending tasks. This, combined with Kotlin's experimental [runBlockingTest](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-test/kotlinx.coroutines.test/run-blocking-test.html) provides a nice way to at least test background tasks that are blocking on coroutines. However, this isn't fool-proof, and has its own issues (e.g. best practice is to use [InstantTaskExecutorRule](https://developer.android.com/reference/android/arch/core/executor/testing/InstantTaskExecutorRule) which actually forces synchronous execution).

Espresso, on the other hand, has the choice of [idling resources](https://developer.android.com/training/testing/espresso/idling-resource). These allow the app to communicate to Espresso when a non-UI bound task has completed which pairs well with Espresso's internal idling mechanism. This helps improve test determinism for Espresso.

Having two solutions isn't viable for shared AndroidX tests, and Robolectric's approach results in synchronizing background processing which reduces the chance of discovering race conditions. Instead, Oppia Android should have a way to run background tasks similar to production (i.e. by not synchronizing them or forcing them to run on the main thread), and ensure both testing frameworks can consistently wait for completion (probably by using idling resource since Robolectric should support them, or using either idling resource or a Robolectric-specific solution). This should also be done in conjunction with running Robolectric in a [paused looper mode](http://robolectric.org/blog/2019/06/04/paused-looper/) to have better parity with Android and try to detect real timing regressions that could occur in production.

Edit: Jetpack Compose also needs to be considered (see comment below).

**Comment by BenHenning** [comment]
This is actually blocked on #59 since this partially requires support for testing libraries in order to have shared testing infrastructure.

**Comment by BenHenning** [comment]
NB: https://github.com/Kotlin/kotlinx.coroutines/issues/242 might be an interesting solution to this if it becomes available.

**Comment by BenHenning** [comment]
NB: this probably blocks exploration fragment tests since the exploration takes a while to load, and we need Espresso to properly idle until all the background processing is complete.

**Comment by BenHenning** [comment]
#927 implements part of this by introducing a test coroutine dispatcher, but that PR is incomplete and still missing a few things:
- Proper Espresso integration (e.g. via an idling resource)
- Proper system clock coordination in test environments
- Proper interruption support (e.g. due to test deadlocks & live locks)
- Preferably, an implementation that doesn't require all tests to declare ``@InternalCoroutinesApi``, though this will require some in-depth investigation to figure out if it's even possible to avoid without reflection

**Comment by BenHenning** [comment]
Reopening since per the comment above this isn't finished yet.

**Comment by BenHenning** [comment]
Also: this probably isn't blocked on #59 since the mechanisms seem to work with pure Gradle.

**Comment by BenHenning** [comment]
See https://github.com/oppia/oppia-android/pull/5825/files#r2246590161. We need to consider Jetpack Compose as well.
